[Variant] Add `variant_get` compute kernel

Samyak2 opened the pull request:
# Which issue does this PR close?

- Closes #7893

# What changes are included in this PR?

In parquet-variant:
- Add a new function `Variant::get_path`: this traverses the path to create a new Variant (does not cast any of it).
- Add a new module `parquet_variant::path`: adds structs/enums to define a path to access a variant value deeply.

In parquet-variant-compute:
- Add a new compute kernel `variant_get`: does the path traversal over a `VariantArray`. In the future, this would also cast the values to a specified type.
- Includes some basic unit tests. Not comprehensive.
- Includes a simple micro-benchmark for reference.

Current limitations:
- It can only return another VariantArray. Casts are not implemented yet.
- Only top-level object/list access is supported. It panics on finding a nested object/list. Needs https://github.com/apache/arrow-rs/pull/7914 to fix this.
- Perf is a TODO.

# Are these changes tested?

Some basic unit tests are added.

# Are there any user-facing changes?

Yes

alamb commented on parquet-variant-compute/src/variant_get.rs:
```diff
@@ -0,0 +1,265 @@
+use crate::utils::variant_from_struct_array;
+
+/// Returns an array with the specified path extracted from the variant values.
+pub fn variant_get(input: &ArrayRef, options: GetOptions) -> Result<ArrayRef> {
```
While trying to review this PR, one thing that I thought might help could be to separate out the Variant traversal and the construction of the output

For example, perhaps you could implement a function to find the appropriate sub Variant  like

```rust
impl Variant {
  fn get_path(&self, path: &VariantPath) -> Option<Variant> {
    ...
  }
}
```

With that building block, you could implement the basic "extract a variant" kernel to a new VariantArray using the newly introduced `VariantArrayBuilder` 

```rust
let mut output_builder = VariantArrayBuilder::new();
// loop over each input row
for input_variant in input_variant_array.iter() {
  // copy the input variant to the output builder
  let mut vb = VariantBuilder::new()
  vb.append(input_variant)
  let (metadata, value) = vb.build();
  output_builder.append_buffers()
}
```

I think once we get https://github.com/apache/arrow-rs/pull/7914 from @friendlymatthew in that should work

One downside of this approach is that it will copy the variant over and if the source and destinations are both BinaryViewArray we can probably be much more clever

alamb replied:
I agree the row based approach will likely be slower for non-shredded variants, but it will always potentially be needed in some cases (for example when the source arrays are not BinaryView)

If we have the `get_path` method, I think we can potentially implement fast copies of variants by playing games with pointers -- basically by checking if the return variant has a pointer into the same buffer of the BinaryViewArray we can make a view that points there.

However, i think that will be somewhat tricky and require some `unsafe` so I suggest we get the first version in plae that does the copy, and once we have it working (and tests written, etc) then I think we'll be in a better position to optimize

Samyak2 replied:
Got it. That makes sense. I'll clean up the PR.

> but it will always potentially be needed in some cases (for example when the source arrays are not BinaryView)

Not sure I follow. Perhaps worth a separate discussion in a follow-up issue.

alamb commented on parquet-variant-compute/src/variant_get.rs:
```diff
@@ -0,0 +1,265 @@
+        vec![true; struct_array.len()]
+    };
+
+    for path in options
```
As I mentioned above I think it might be easier to implement this function as a method on variant itself

alamb commented on parquet-variant-compute/src/variant_get.rs:
```diff
@@ -0,0 +1,265 @@
+        }
+    }
+
+    let as_type = options.as_type.ok_or_else(|| {
```
another way to potentially simplify / break this PR up into smaller parts would be to first focus on a `variant_get` that only supports returning the requested field as another VariantArray.

We could then work out how to support casting to primitive types after that.

Samyak2 replied:
That makes sense. I'll focus on returning a variant for this PR

alamb reviewed (commented):
Thanks @Samyak2  -- this is looking like a great start. I left some ideas on how to break it up / make progress with smaller PRs. 

fyi @scovich and @friendlymatthew

scovich commented on parquet-variant-compute/src/utils.rs:
```diff
@@ -0,0 +1,48 @@
+        }
+    }
+
+    let metadata_array = struct_array
+        .column(0)
+        .as_any()
+        .downcast_ref::<BinaryArray>()
+        .ok_or_else(|| ArrowError::CastError("Expected BinaryArray for 'metadata'".into()))?;
+
+    let value_array = struct_array
+        .column(1)
```
Does the `VariantArray ` guarantee that metadata is the first column? 
Because the parquet spec specifically states that one must rely on field names, not ordinals.

Samyak2 replied:
Yep, you're right. I copied this code from `batch_variant_to_json_string`. Now that we have VariantArray, we don't need this anymore.

alamb reviewed (approved):
Thank you @Samyak2  -- I think this is a great first step. I agree there are a bunch of TODOs but I also think we could merge this as is and then iterate on the output

I am curious about your thoughts on the API design @scovich and @friendlymatthew 

cc @PinkCrow007 @mprammer

Samyak2 commented:
Made the following changes:
- Moved to a non-owned Cow string like @alamb originally suggested in the issue
- Added more tests
- Fixed formatting and missing license header issues

scovich commented on parquet-variant-compute/src/variant_get.rs:
```diff
@@ -0,0 +1,202 @@
+    for i in 0..variant_array.len() {
+        let new_variant = variant_array.value(i);
+        // TODO: perf?
+        let new_variant = new_variant.get_path(&options.path);
+        if let Some(new_variant) = new_variant {
+            // TODO: we're decoding the value and doing a copy into a variant value again. This
+            // copy can be much smarter.
+            builder.append_variant(new_variant);
```
Can't we just do a directly byte-slice copy? The input is either fully validated or the caller has asserted that the bytes are correct. Maybe a `VariantArray::value_bytes` method and then `VariantArrayBuilder::append_variant_bytes`? Hopefully, that would also (eventually) map more nicely to a view-based approach.

Samyak2 replied:
That would be a better way, but I don't think that's possible here? `new_variant` here is of type `Variant`, which is a decoded variant value. So it doesn't really retain the raw bytes of the metadata/value from which it was constructed. It does retain the metadata and value bytes in case of objects/lists, but that case isn't working now anyway.

One way would be to operate on metadata/value byte slices directly, even for the `get_path` operation. That way, we'll be left with a `value` byte slice that we can copy directly.

scovich commented on parquet-variant/src/path.rs:
```diff
@@ -0,0 +1,64 @@
+}
+
+impl<'a> Deref for VariantPath<'a> {
+    type Target = Vec<VariantPathElement<'a>>;
```
Normally `Deref` should go "all the way" and invoke Deref for the target type as well?
Which in this case would mean `Target = [VariantPathElement<'a>]`?

scovich commented on parquet-variant/src/variant.rs:
```diff
@@ -1063,6 +1064,34 @@ impl<'m, 'v> Variant<'m, 'v> {
+    pub fn get_path(&self, path: &VariantPath) -> Option<Variant> {
+        let mut output = self.clone();
+        for element in path.deref() {
+            match element {
+                VariantPathElement::Field { name } => {
+                    let Variant::Object(variant_object) = output else {
+                        return None;
+                    };
+                    let field = variant_object.get(name);
+                    let field = field?;
+                    output = field;
+                }
+                VariantPathElement::Index { index } => {
+                    let Variant::List(variant_list) = output else {
+                        return None;
+                    };
+                    let index = variant_list.get(*index);
+                    let index = index?;
+                    output = index;
+                }
+            }
```
```suggestion
            output = match element {
                VariantPathElement::Field { name } => output.get_object_field(name)?,
                VariantPathElement::Index { index } => output.get_array_element(*index)?,
            };
```

Samyak2 replied:
That is so much cleaner. Thanks!

scovich commented on parquet-variant/src/variant.rs:
```diff
@@ -1063,6 +1064,34 @@ impl<'m, 'v> Variant<'m, 'v> {
+    /// If the path is not found, `None` is returned.
+    pub fn get_path(&self, path: &VariantPath) -> Option<Variant> {
+        let mut output = self.clone();
+        for element in path.deref() {
```
It's not usually necessary to invoke `deref` directly?

scovich replied:
... but apparently this is a context where deref isn't automatic (tried it in playground). Tho we could perhaps be more clear by 
```suggestion
        for element in path.iter() {
```
... which _does_ deref automatically?

Samyak2 commented on parquet-variant-compute/src/variant_get.rs:
```diff
@@ -0,0 +1,202 @@
+    pub path: VariantPath<'a>,
+    /// if `as_type` is None, the returned array will itself be a VariantArray.
+    ///
+    /// if `as_type` is `Some(type)` the field is returned as the specified type if possible. To specify returning
+    /// a Variant, pass a Field with variant type in the metadata.
+    pub as_type: Option<Field>,
```
> But meanwhile -- I tend to agree with @alamb that we should separate out casting from pathing. If get_path anyway has to return a Variant, somebody (ie arrow array builder) who actually cares about the type will still have to match on the result even if they just panic in the "impossible" case of an unexpected enum variant. Seems better to let the caller do the casting themselves since they probably have a strongly-typed place to put the result.

This makes sense. But I'm not sure what you're suggesting here? Should we have a version of `variant_get` here that doesn't do casting and always returns a `VariantArray`?

> As a bonus, that would also mean we don't need to define GetOptions at this low level -- get_path only needs to know about VariantPath. Arrow variant compute and the array builders could worry about the rest of the options space.

`Variant::get_path` already only knows about `VariantPath`. `GetOptions` is defined in this `parquet-variant-compute` crate. Or are you suggesting that we add a `get_path` method to `VariantArray`?

alamb commented:
I merged up from main and fixed a logical conflict. Thanks again @Samyak2  -- once this PR passes CI I'll merge it in

alamb commented:
🚀

alamb commented:
Thanks again @Samyak2  -- this is a great initial addtion

alamb commented:
I made a small PR here:
 -https://github.com/apache/arrow-rs/pull/7942

alamb commented:
And it seems @carpecodeum  is already using this PR as a base 🥳 
- https://github.com/cmu-db/arrow-rs/pull/2
